lib/crypto.js

```diff
@@ -0,0 +1,48 @@
+const sodium = require('sodium-universal')
+const b4a = require('b4a')
+const c = require('compact-encoding')
+
+exports.data = function (data) {
+  const out = b4a.allocUnsafe(32)
+
+  sodium.crypto_generichash_batch(out, [
+    c.encode(c.uint64, data.byteLength),
+    data
+  ])
+
+  return out
+}
+
+exports.parent = function (a, b) {
+  if (a.index > b.index) {
+    const tmp = a
+    a = b
+    b = tmp
+  }
+
+  const out = b4a.allocUnsafe(32)
+
+  sodium.crypto_generichash_batch(out, [
+    c.encode(c.uint64, a.size + b.size),
+    a.hash,
+    b.hash
+  ])
+
+  return out
+}
+
+exports.tree = function (roots, out) {
+  const buffers = new Array(3 * roots.length)
+  let j = 0
+
+  for (let i = 0; i < roots.length; i++) {
+    const r = roots[i]
+    buffers[j++] = r.hash
+    buffers[j++] = c.encode(c.uint64, r.index)
+    buffers[j++] = c.encode(c.uint64, r.size)
+  }
+
+  if (!out) out = b4a.allocUnsafe(32)
+  sodium.crypto_generichash_batch(out, buffers)
+  return out
+}
```

Comment:
fyi these are lifted from hypercore-crypto, just with the id byte removed. vendored here so i can support the legacy flag